### ISSUE TITLE
Read the docs builds are broken

### ISSUE THREAD

**Issue by weaverba137** [desihub/redrock#80]
Subsequent to enabling Read the Docs for redrock, the documentation builds on the master branch were broken.  I think this is due to unnecessary imports in the top-level `__init__.py` file.

**Comment by tskisner** [comment]
The sphinx build succeeds, both locally and on travis.  What other steps does read the docs require?  Just wanted to know so that I can check that in the future...

**Comment by weaverba137** [comment]
The Read the Docs environment is very "bare" compared to Travis.  You can assume that packages related to Sphinx are installed, but you can't assume that any numerical packages, such as `numpy` are installed.

**Comment by weaverba137** [comment]
PS, you can also look at the badge on the README.rst file.  Right now that says Read the Docs is failing.

**Comment by weaverba137** [comment]
Also, please let me know if you can view this link: http://readthedocs.org/projects/redrock/builds/

**Comment by tskisner** [comment]
Hi @weaverba137 , yes I can see that status page on readthedocs.  Frequently I use __init__.py as a way of controlling the "commonly used" functions / classes with relative imports of specific classes from submodules.  So one can do "from foo import Foo", rather than "from foo.foo import Foo".  However, in larger packages with mostly independent sub-modules this kind of namespace convenience doesn't work well.

Obviously if those convenience imports in __init__.py are causing problems then we should get rid of them.

**Comment by weaverba137** [comment]
To drill down into the technical details, here's what happens:

1. Sphinx starts up, using the `sphinx-build` command, rather than `python setup.py build_sphinx`.
2. Sphinx reads `doc/conf.py`.
3. `doc/conf.py` imports `py/redrock/__init__.py` to get the version string.
4. `py/redrock/__init__.py` performs a whole bunch of other imports, and some of those modules try to import modules that are not present.
5. **However** if the import of the version string succeeds (*i.e.* we fix `__init__.py` so it doesn't perform those problematic imports), then `doc/conf.py` goes on to define *mock* modules to satisfy dependencies, so the various redrock modules can be imported successfully.
6. The documentation builds successfully.

In other words, you *can* build the documentation in a bare environment, but you can't have any imports that might fail and prevent reading the `__version__` string.

I'm still waiting for Read the Docs to detect the branch, so I can make sure it succeeds.

**Comment by sbailey** [comment]
Do we have any other options for configuring the readthedocs build method to work around this?  It seems like the burden should be on the readthedocs build to work with our code, rather than having to change how we want to import our packages just so that they can be built on readthedocs.

**Comment by weaverba137** [comment]
There are several possible options. The fundamental thing to keep in mind is that *Sphinx*, not just ReadTheDocs, needs to know the version of the software.  Also keep in mind that ReadTheDocs uses the `sphinx-build` command instead of `python setup.py build_sphinx`.  Here are some ways that Sphinx can figure out the version of the software:

1. Import `redrock.__version__`.  This is what it does now.
2. Use `desiutil.setup.get_version()`.  This is what the typical DESI `setup.py` file uses, although `get_version()` is specialized to work in the same directory that contains the `setup.py` file.
3. Hard-code the version string into the `doc/conf.py` file.  This could include modifying `desiutil.setup.DesiVersion` so that when it updates `redrock._version`, it also updates the `doc/conf.py` file.

If we want to stick with item 1, then here are some ways in which that could be made to work:

1. Remove unnecessary imports from `__init__.py`.  I've already done the work to get the builds working again on the `fix-rtd-builds` branch.
2. Move all imports of external packages inside the function definitions that need them.
3. There is experimental support for building docs within a conda environment.  See https://docs.readthedocs.io/en/latest/yaml-config.html for more details.  This will require more work, and it would also require installing desispec in the ReadTheDocs environment.

For reference here are the imports that need to succeed when Sphinx builds documentation (and this is what it needs to actually read the docstrings, *after* it has successfully determined the version string):
```python
autodoc_mock_imports = ['astropy', 'astropy.io', 'astropy.table',
                        'desispec', 'desispec.io', 'desispec.resolution',
                        'fitsio', 'numba', 'numpy',
                        'scipy.constants', 'scipy.sparse'] 
```

Now, for my *opinion*. As @tskisner says, these imports only exist for convenience: they provide *aliases* for classes and functions that could equally well be imported directly from the module they reside in.  They are not necessary to the functioning of the package, and it's not like our API is secret or otherwise needs to be hidden somehow.  In my opinion, these imports cause more trouble in terms of imports happening at unexpected times than any convenience they provide.  I would be interested to hear if there are other reasons, besides convenience, for setting up `__init__.py` files this way.

**Comment by weaverba137** [comment]
PS, the branch `fix-rtd-builds` is building successfully: http://redrock.readthedocs.io/en/fix-rtd-builds/

**Comment by tskisner** [comment]
Just to add to @weaverba137's statement.  All python files inside the redrock package use relative imports (which are much safer, and guarantee that we are importing files from within the same install prefix or source tree).

So the "convenience" imports in ```__init__.py``` are only there for external scripts using the redrock module directly.  Removing them from the top-level init should have no impact on the internal workings of the package (if it does, that's a bug).

Let's just remove them.  Anyone importing redrock in a debug script or python notebook can update their import statements as needed.

**Comment by sbailey** [comment]
Thanks for itemizing the options, @weaverba137 .  Your PR #91 (merged) fixes this for now via option 1 (import redrock.__version__ and restrict what kind of other imports we can include so that it doesn't break RTD).  Longer term option (2) seems preferable (use desiutil.setup.get_version).  It's ok for redrock to have an optional dependency on desiutil, and setup.py already has that, and our RTD config could use that fact (right? or am I missing something?)

Although that other style of import is "only" for convenience, I think convenience is a serious consideration for developer productivity so let's not forbid that if we don't have to.

**Comment by weaverba137** [comment]
To be sure, I'm not opposed to convenience, but if a random developer wants that convenience, there are also consequences.  For example, go ahead & include those imports in `__init__.py` but make certain that any imports of external dependencies are inside function definitions, not module-level.
